fineract-provider/src/main/java/org/apache/fineract/adhocquery/service/AdHocWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -0,0 +1,197 @@
+    private final PlatformSecurityContext context;
+    private final AdHocRepository adHocRepository;
+    private final AdHocDataValidator adHocCommandFromApiJsonDeserializer;
+   
+
+    @Autowired
+    public AdHocWritePlatformServiceJpaRepositoryImpl(final PlatformSecurityContext context, final AdHocRepository roleRepository,
+             final AdHocDataValidator adHocCommandFromApiJsonDeserializer) {
+        this.context = context;
+        this.adHocRepository = roleRepository;
+        this.adHocCommandFromApiJsonDeserializer = adHocCommandFromApiJsonDeserializer;
+       
+    }
+
+    @Transactional
+    @Override
+    public CommandProcessingResult createAdHocQuery(final JsonCommand command) {
+
+        try {
+            this.context.authenticatedUser();
+
+            this.adHocCommandFromApiJsonDeserializer.validateForCreate(command.json());
+
+            final AdHoc entity = AdHoc.fromJson(command);
+            this.adHocRepository.save(entity);
+
+            return new CommandProcessingResultBuilder().withCommandId(command.commandId()).withEntityId(entity.getId()).build();
+        } catch (final DataIntegrityViolationException dve) {
+            handleDataIntegrityIssues(command, dve);
+            return new CommandProcessingResultBuilder() //
+                    .withCommandId(command.commandId()) //
+                    .build();
+        }
+    }
+
+    /*
+     * Guaranteed to throw an exception no matter what the data integrity issue
+     * is.
+     */
+    private void handleDataIntegrityIssues(final JsonCommand command, final DataIntegrityViolationException dve) {
+
+        final Throwable realCause = dve.getMostSpecificCause();
+        if (realCause.getMessage().contains("unq_name")) {
+
+            final String name = command.stringValueOfParameterNamed("name");
+            throw new PlatformDataIntegrityException("error.msg.role.duplicate.name", "Role with name `" + name + "` already exists",
+                    "name", name);
+        }
+
+        logAsErrorUnexpectedDataIntegrityException(dve);
+        throw new PlatformDataIntegrityException("error.msg.role.unknown.data.integrity.issue",
+                "Unknown data integrity issue with resource.");
+    }
+
+    private void logAsErrorUnexpectedDataIntegrityException(final DataIntegrityViolationException dve) {
+        logger.error(dve.getMessage(), dve);
+    }
+
+    //@Caching(evict = { @CacheEvict(value = "users", allEntries = true), @CacheEvict(value = "usersByUsername", allEntries = true) })
+    @Transactional
+    @Override
+    public CommandProcessingResult updateAdHocQuery(final Long roleId, final JsonCommand command) {
+        try {
+            this.context.authenticatedUser();
+
+            this.adHocCommandFromApiJsonDeserializer.validateForUpdate(command.json());
+
+            final AdHoc adHoc = this.adHocRepository.findOne(roleId);
+            if (adHoc == null) { throw new RoleNotFoundException(roleId); }
```

Comment:
if adHoc is  null you should throw AdHocNotFoundException  not RoleNotFoundException?